compiler/rustc_trait_selection/src/traits/const_evaluatable.rs

```diff
@@ -619,32 +622,59 @@ where
     recurse(tcx, ct, &mut f)
 }
 
+// Substitutes generics repeatedly to allow AbstractConsts to unify where a
+// ConstKind::Unevalated could be turned into an AbstractConst that would unify e.g.
+// Param(N) should unify with Param(T), substs: [Unevaluated("T2", [Unevaluated("T3", [Param(N)])])]
+#[inline]
+#[instrument(skip(tcx), level = "debug")]
+fn try_replace_substs_in_root<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    mut abstr_const: AbstractConst<'tcx>,
+) -> Option<AbstractConst<'tcx>> {
+    while let Node::Leaf(ct) = abstr_const.root(tcx) {
+        match AbstractConst::from_const(tcx, ct) {
+            Ok(Some(act)) => abstr_const = act,
+            Ok(None) => break,
+            Err(_) => return None,
+        }
+    }
+
+    Some(abstr_const)
+}
+
 /// Tries to unify two abstract constants using structural equality.
+#[instrument(skip(tcx), level = "debug")]
 pub(super) fn try_unify<'tcx>(
```

Comment:
we may want to make this a method on a `ConstUnifyContext` or whatever, considering that we have to pass `tcx` and `param_env` through for every method call